// webpack-plugin-critical/dist/critical.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var path_1 = require("path");
var critical = require('critical');
var webpack_util_cleantmp_1 = require("webpack-util-cleantmp");
var CriticalPlugin = (function () {
    function CriticalPlugin(options) {
        this.options = options;
        if (!options.dest) {
            throw new Error('A `dest` option is required for the Critical webpack plugin');
        }
    }
    CriticalPlugin.prototype.emit = function (compilation, callback) {
        var options = this.options;
        var subscription = webpack_util_cleantmp_1.cleantmp({
            prefix: 'criticalcss',
            globToDisk: '**/*.+(css|html)',
            globFromDisk: '**/*.+(css|html)',
            assets: compilation.assets
        })
            .subscribe(function (tmp) {
            var opts = __assign({}, options, {
                base: path_1.resolve(tmp, options.base || ''),
                dest: path_1.resolve(tmp, options.dest)
            });
            critical.generate(opts, function (err, output) {
                subscription.unsubscribe();
                callback(err);
            });
        });
    };
    CriticalPlugin.prototype.apply = function (compiler) {
        compiler.plugin("emit", this.emit.bind(this));
    };
    return CriticalPlugin;
}());
exports.CriticalPlugin = CriticalPlugin;
